Multi-bug discovery during generation

As a consequence of #836 we can now find and display multiple bugs if these occur during either shrinking or retrieving examples from the database.

Amusingly though, we can't currently find multiple bugs in the phase that is actually designed to find bugs, generation! We should probably fix that.

However it is not sufficient to just run generation to completion, because a reasonable usage pattern is "Run generation until you find a bug and then stop and tell me that bug", which could be done by e.g. setting `max_examples` absurdly high and turning off the time limit. In this case we would not want to keep running indefinitely after we've found a bug.

I think the correct logic here is:

1. Skip generation altogether if we found a bug in example replay
2. Always respect `max_examples` in the sense that we stop running generation after we've generated that many valid or interesting examples.
3. Stop running generation at some point "reasonably soon" after we've found our first bug.

The devil is of course in the details of what "reasonably soon" means. A non-exhaustive list of possibilities:

1. Just do it time based - run for at most a few seconds after the first bug is found.
2. Stop when we've run twice as many examples as it took to find the first bug
3. Stop when we've run twice as many examples as it took to find the *last* bug
4. Do something clever and statistical to estimate the number of bugs we can expect to find and stop when we've found that many
5. Some hybrid of the above.

My preference is probably some hybrid of 1 and 3.

To make the point, we should also update our headline example in the readme:

```python
@given(st.lists(st.floats(allow_nan=False)))
def test_mean(xs):
    mean = sum(xs) / len(xs)
    assert min(xs) <= mean <= max(xs)
```
```
Falsifying example: test_mean(xs=[9.9792015476736e+291, 1.7976931348623157e+308])
Traceback (most recent call last):
    ...
AssertionError: mean=inf

Falsifying example: test_mean(xs=[])
Traceback (most recent call last):
    ...
ZeroDivisionError: division by zero

---------------------------------------------------------------------------
MultipleFailures: Hypothesis found 2 distinct failures.
```

(Note that this is on the current released version!  We already find two distinct bugs, just by shrinking the list 😄)

--- a/hypothesis-python/RELEASE.rst
+++ b/hypothesis-python/RELEASE.rst
@@ -1,0 +1,6 @@
+RELEASE_TYPE: patch
+
+This patch allows Hypothesis to try a few more examples after finding the
+first bug, in hopes of reporting multiple distinct bugs.  The heuristics
+described in :issue:`847` ensure that we avoid wasting time on fruitless
+searches, while still surfacing each bug as soon as possible.

--- a/hypothesis-python/src/hypothesis/internal/conjecture/engine.py
+++ b/hypothesis-python/src/hypothesis/internal/conjecture/engine.py
@@ -100,8 +100,8 @@ class ConjectureRunner(object):
 
         self.interesting_examples = {}
         # We use call_count because there may be few possible valid_examples.
-        self.first_bug_calls = 0
-        self.last_bug_calls = 0
+        self.first_bug_found_at = None
+        self.last_bug_found_at = None
 
         self.shrunk_examples = set()
 
@@ -186,9 +186,9 @@ class ConjectureRunner(object):
                 existing = self.interesting_examples[key]
             except KeyError:
                 changed = True
-                self.last_bug_calls = self.call_count
-                if not self.first_bug_calls:
-                    self.first_bug_calls = self.call_count
+                self.last_bug_found_at = self.call_count
+                if self.first_bug_found_at is None:
+                    self.first_bug_found_at = self.call_count
             else:
                 if sort_key(data.buffer) < sort_key(existing.buffer):
                     self.shrinks += 1
@@ -206,8 +206,10 @@ class ConjectureRunner(object):
                 self.exit_with(ExitReason.max_shrinks)
 
         if not self.interesting_examples:
-            # Note that this logic is reproduced to end the generation phase when we
-            # have interesting examples.  Update that too if you change this!
+            # Note that this logic is reproduced to end the generation phase when
+            # we have interesting examples.  Update that too if you change this!
+            # (The doubled implementation is because here we exit the engine entirely,
+            #  while in the other case below we just want to move on to shrinking.)
             if self.valid_examples >= self.settings.max_examples:
                 self.exit_with(ExitReason.max_examples)
             if self.call_count >= max(
@@ -654,23 +656,26 @@ class ConjectureRunner(object):
             # doing anything else as it's by far the most common case.
             if not self.interesting_examples:
                 return True
-            assert self.first_bug_calls <= self.last_bug_calls <= self.call_count
-            # End the generation phase where we would have ended it if no bugs
-            # had been found.  Reproduces the exit logic in `self.test_case`.
+            # If we've found a bug and won't report more than one, stop looking.
+            elif not self.settings.report_multiple_bugs:
+                return False
+            assert self.first_bug_found_at <= self.last_bug_found_at <= self.call_count
+            # End the generation phase where we would have ended it if no bugs had
+            # been found.  This reproduces the exit logic in `self.test_function`,
+            # but with the important distinction that this clause will move on to
+            # the shrinking phase having found one or more bugs, while the other
+            # will exit having found zero bugs.
             if (
                 self.valid_examples >= self.settings.max_examples
                 or self.call_count >= max(self.settings.max_examples * 10, 1000)
             ):  # pragma: no cover
-                return False
-            # If we've found a bug and won't report more than one, stop looking.
-            if self.interesting_examples and not self.settings.report_multiple_bugs:
                 return False
             # Otherwise, keep searching for between ten and 'a heuristic' calls.
             # We cap 'calls after first bug' so errors are reported reasonably
             # soon even for tests that are allowed to run for a very long time,
             # or sooner if the latest half of our test effort has been fruitless.
             return self.call_count < MIN_TEST_CALLS or self.call_count < min(
-                self.first_bug_calls + 1000, self.last_bug_calls * 2
+                self.first_bug_found_at + 1000, self.last_bug_found_at * 2
             )
 
         count = 0

--- a/hypothesis-python/src/hypothesis/internal/conjecture/engine.py
+++ b/hypothesis-python/src/hypothesis/internal/conjecture/engine.py
@@ -56,6 +56,7 @@ __tracebackhide__ = True
 MAX_SHRINKS = 500
 CACHE_SIZE = 10000
 MUTATION_POOL_SIZE = 100
+MIN_TEST_CALLS = 10
 
 
 @attr.s
@@ -98,6 +99,9 @@ class ConjectureRunner(object):
         self.target_selector = TargetSelector(self.random)
 
         self.interesting_examples = {}
+        # We use call_count because there may be few possible valid_examples.
+        self.first_bug_calls = 0
+        self.last_bug_calls = 0
 
         self.shrunk_examples = set()
 
@@ -182,6 +186,9 @@ class ConjectureRunner(object):
                 existing = self.interesting_examples[key]
             except KeyError:
                 changed = True
+                self.last_bug_calls = self.call_count
+                if not self.first_bug_calls:
+                    self.first_bug_calls = self.call_count
             else:
                 if sort_key(data.buffer) < sort_key(existing.buffer):
                     self.shrinks += 1
@@ -199,6 +206,8 @@ class ConjectureRunner(object):
                 self.exit_with(ExitReason.max_shrinks)
 
         if not self.interesting_examples:
+            # Note that this logic is reproduced to end the generation phase when we
+            # have interesting examples.  Update that too if you change this!
             if self.valid_examples >= self.settings.max_examples:
                 self.exit_with(ExitReason.max_examples)
             if self.call_count >= max(
@@ -588,6 +597,11 @@ class ConjectureRunner(object):
     def generate_new_examples(self):
         if Phase.generate not in self.settings.phases:
             return
+        if self.interesting_examples:
+            # The example database has failing examples from a previous run,
+            # so we'd rather report that they're still failing ASAP than take
+            # the time to look for additional failures.
+            return
 
         zero_data = self.cached_test_function(hbytes(self.settings.buffer_size))
         if zero_data.status > Status.OVERRUN:
@@ -635,8 +649,32 @@ class ConjectureRunner(object):
 
         self.health_check_state = HealthCheckState()
 
+        def should_generate_more():
+            # If we haven't found a bug, keep looking.  We check this before
+            # doing anything else as it's by far the most common case.
+            if not self.interesting_examples:
+                return True
+            assert self.first_bug_calls <= self.last_bug_calls <= self.call_count
+            # End the generation phase where we would have ended it if no bugs
+            # had been found.  Reproduces the exit logic in `self.test_case`.
+            if (
+                self.valid_examples >= self.settings.max_examples
+                or self.call_count >= max(self.settings.max_examples * 10, 1000)
+            ):  # pragma: no cover
+                return False
+            # If we've found a bug and won't report more than one, stop looking.
+            if self.interesting_examples and not self.settings.report_multiple_bugs:
+                return False
+            # Otherwise, keep searching for between ten and 'a heuristic' calls.
+            # We cap 'calls after first bug' so errors are reported reasonably
+            # soon even for tests that are allowed to run for a very long time,
+            # or sooner if the latest half of our test effort has been fruitless.
+            return self.call_count < MIN_TEST_CALLS or self.call_count < min(
+                self.first_bug_calls + 1000, self.last_bug_calls * 2
+            )
+
         count = 0
-        while not self.interesting_examples and (
+        while should_generate_more() and (
             count < 10 or self.health_check_state is not None
         ):
             prefix = self.generate_novel_prefix()
@@ -661,7 +699,7 @@ class ConjectureRunner(object):
 
         zero_bound_queue = []
 
-        while not self.interesting_examples:
+        while should_generate_more():
             if zero_bound_queue:
                 # Whenever we generated an example and it hits a bound
                 # which forces zero blocks into it, this creates a weird

--- a/hypothesis-python/tests/cover/test_conjecture_engine.py
+++ b/hypothesis-python/tests/cover/test_conjecture_engine.py
@@ -36,6 +36,7 @@ from hypothesis.internal.conjecture.data import (
     Status,
 )
 from hypothesis.internal.conjecture.engine import (
+    MIN_TEST_CALLS,
     ConjectureRunner,
     ExitReason,
     TargetSelector,
@@ -191,7 +192,7 @@ def test_detects_flakiness():
     runner = ConjectureRunner(tf)
     runner.run()
     assert runner.exit_reason == ExitReason.flaky
-    assert count == [2]
+    assert count == [MIN_TEST_CALLS + 1]
 
 
 def test_variadic_draw():
@@ -271,6 +272,32 @@ def test_stops_after_max_examples_when_generating():
     assert len(seen) == 1
 
 
+@pytest.mark.parametrize("examples", [1, 5, 20, 50])
+def test_stops_after_max_examples_when_generating_more_bugs(examples):
+    seen = []
+    bad = [False, False]
+
+    def f(data):
+        seen.append(data.draw_bits(32))
+        # Rare, potentially multi-error conditions
+        if seen[-1] > 2 ** 31:
+            bad[0] = True
+            raise ValueError
+        bad[1] = True
+        raise Exception
+
+    runner = ConjectureRunner(
+        f, settings=settings(max_examples=examples, phases=[Phase.generate])
+    )
+    try:
+        runner.run()
+    except Exception:
+        pass
+    # No matter what, whether examples is larger or smalller than MAX_TEST_CALLS,
+    # we stop looking at max_examples.  (and re-run each failure for the traceback)
+    assert len(seen) <= examples + sum(bad)
+
+
 def test_interleaving_engines():
     children = []
 
@@ -304,7 +331,7 @@ def test_phases_can_disable_shrinking():
         f, settings=settings(database=None, phases=(Phase.reuse, Phase.generate))
     )
     runner.run()
-    assert len(seen) == 1
+    assert len(seen) == MIN_TEST_CALLS
 
 
 def test_erratic_draws():

--- a/hypothesis-python/tests/cover/test_flakiness.py
+++ b/hypothesis-python/tests/cover/test_flakiness.py
@@ -21,6 +21,7 @@ import pytest
 
 from hypothesis import HealthCheck, Verbosity, assume, example, given, reject, settings
 from hypothesis.errors import Flaky, Unsatisfiable, UnsatisfiedAssumption
+from hypothesis.internal.conjecture.engine import MIN_TEST_CALLS
 from hypothesis.strategies import booleans, composite, integers, lists, random_module
 from tests.common.utils import no_shrink
 
@@ -59,6 +60,7 @@ def test_gives_flaky_error_if_assumption_is_flaky():
 def test_does_not_attempt_to_shrink_flaky_errors():
     values = []
 
+    @settings(database=None)
     @given(integers())
     def test(x):
         values.append(x)
@@ -66,7 +68,12 @@ def test_does_not_attempt_to_shrink_flaky_errors():
 
     with pytest.raises(Flaky):
         test()
-    assert len(set(values)) == 1
+    # We try a total of ten calls in the generation phase, each usually a
+    # unique value, looking briefly (and unsuccessfully) for another bug.
+    assert 1 < len(set(values)) <= MIN_TEST_CALLS
+    # We don't try any new values while shrinking, just execute the test
+    # twice more (to check for flakiness and to raise the bug to the user).
+    assert set(values) == set(values[:-2])
 
 
 class SatisfyMe(Exception):

--- a/hypothesis-python/tests/cover/test_slippage.py
+++ b/hypothesis-python/tests/cover/test_slippage.py
@@ -20,9 +20,10 @@ from __future__ import absolute_import, division, print_function
 import pytest
 
 import hypothesis.strategies as st
-from hypothesis import given, settings
+from hypothesis import Phase, given, settings
 from hypothesis.database import InMemoryExampleDatabase
 from hypothesis.errors import Flaky, MultipleFailures
+from hypothesis.internal.conjecture.engine import MIN_TEST_CALLS
 from tests.common.utils import capture_out, non_covering_examples
 
 
@@ -239,3 +240,51 @@ def test_can_disable_multiple_error_reporting(allow_multi):
     with pytest.raises(MultipleFailures if allow_multi else TypeError):
         test()
     assert seen == {TypeError, ValueError}
+
+
+def test_finds_multiple_failures_in_generation():
+    @settings(phases=[Phase.generate])
+    @given(st.lists(st.floats()))
+    def test(x):
+        mean = sum(x) / len(x)  # ZeroDivisionError
+        if len(x) == 1:
+            assert mean == x[0]  # x[0] is nan
+        assert min(x) <= mean <= max(x)  # x contains nan or +&- inf
+
+    with pytest.raises(MultipleFailures):
+        test()
+
+
+def test_stops_immediately_if_not_report_multiple_bugs():
+    seen = set()
+
+    @settings(phases=[Phase.generate], report_multiple_bugs=False)
+    @given(st.integers())
+    def test(x):
+        seen.add(x)
+        assert False
+
+    with pytest.raises(AssertionError):
+        test()
+    assert len(seen) == 1
+
+
+def test_stops_immediately_on_replay():
+    seen = set()
+
+    @settings(database=InMemoryExampleDatabase())
+    @given(st.integers())
+    def test(x):
+        seen.add(x)
+        assert x
+
+    # On the first run, we look for up to ten examples:
+    with pytest.raises(AssertionError):
+        test()
+    assert 1 < len(seen) <= MIN_TEST_CALLS
+
+    # With failing examples in the database, we stop at one.
+    seen.clear()
+    with pytest.raises(AssertionError):
+        test()
+    assert len(seen) == 1

--- a/hypothesis-python/tests/cover/test_testdecorators.py
+++ b/hypothesis-python/tests/cover/test_testdecorators.py
@@ -23,6 +23,7 @@ from collections import namedtuple
 
 import hypothesis.reporting as reporting
 from hypothesis import HealthCheck, Verbosity, assume, given, note, settings
+from hypothesis.internal.conjecture.engine import MIN_TEST_CALLS
 from hypothesis.strategies import (
     binary,
     booleans,
@@ -417,7 +418,7 @@ def test_when_set_to_no_simplifies_runs_failing_example_twice():
     with raises(AssertionError):
         with capture_out() as out:
             foo()
-    assert failing == [2]
+    assert failing[0] <= MIN_TEST_CALLS + 2
     assert "Falsifying example" in out.getvalue()
     assert "Lo" in out.getvalue()
 

